src/lib/comms/sol-oic-security.c

```diff
@@ -53,34 +51,51 @@ struct cred_item {
     /* FIXME: Only symmetric pairwise keys supported at the moment. */
 };
 
-struct creds {
-    struct sol_vector items;
-    const struct sol_oic_security *security;
-};
-
 struct sol_socket *sol_coap_server_get_socket(const struct sol_coap_server *server);
 
 static ssize_t
 creds_get_psk(const void *data, struct sol_str_slice id,
     char *psk, size_t psk_len)
 {
-    const struct creds *creds = data;
-    struct cred_item *iter;
-    uint16_t idx;
+    SOL_BUFFER_DECLARE_STATIC(buf, PATH_MAX);
+    struct sol_cert *cert;
+    struct sol_blob *contents;
+    int r;
 
     SOL_DBG("Looking for PSK with ID=%.*s", (int)id.len, id.data);
 
-    SOL_VECTOR_FOREACH_IDX (&creds->items, iter, idx) {
-        if (sol_str_slice_eq(id, iter->id.slice)) {
-            if (iter->id.slice.len > psk_len)
-                return -ENOBUFS;
+    r = sol_buffer_append_printf(&buf, "%s", OIC_CRED_FILE_PREFIX);
+    SOL_INT_CHECK(r, < 0, -ENOBUFS);
+    r = sol_buffer_append_as_base64(&buf, id, SOL_BASE64_MAP);
+    SOL_INT_CHECK(r, < 0, -ENOBUFS);
+    r = sol_buffer_append_printf(&buf, "%s", OIC_CRED_FILE_SUFIX);
+    SOL_INT_CHECK(r, < 0, -ENOBUFS);
+
+    cert = sol_cert_load_from_file(buf.data);
+    if (!cert)
+        return -ENOENT;
 
-            memcpy(psk, iter->psk.data, iter->id.slice.len);
-            return (ssize_t)iter->psk.slice.len;
-        }
+    sol_buffer_reset(&buf);
+
+    contents = sol_cert_get_contents(cert);
+    sol_cert_unref(cert);
+
+    if (contents->size < PSK_BASE64_LEN) {
+        sol_blob_unref(contents);
+        return -ENOENT;
     }
 
-    return -ENOENT;
+    r = sol_buffer_append_from_base64(&buf,
+        SOL_STR_SLICE_STR(contents->mem, PSK_BASE64_LEN), SOL_BASE64_MAP);
+    sol_blob_unref(contents);
+    SOL_INT_CHECK(r, < 0, -ENOBUFS);
+
+    if (buf.used > psk_len)
+        return -ENOBUFS;
+
+    memcpy(psk, buf.data, buf.used);
```

Comment:
why are you storing as base64 and decoding it here?

Also, please keep the blob as further as possible to reduce memory pressure. On Linux, the blob will `mmap()`, on small OSes they can point directly to the storage in flash or RAM. Duplicating is bad.